Fairmat 2024: specialized base classes on top of NXprocess

lukaspie opened the pull request:
Adds several new base classes that all extend `NXprocess`:
- `NXcalibration`: for describing calibration of an instrument and/or a data set (e.g., an energy axis) -> moved from contributed to base_classes
- `NXdistortion`: post-processing distortion corrections
- `NXregistration`: [image registration](https://en.wikipedia.org/wiki/Image_registration)

phyy-nx commented on base_classes/NXprocess.nxdl.xml:
```diff
@@ -43,6 +43,21 @@
+        <doc>
+             Describes the operations of calibration procedures, e.g. axis calibrations.
+        </doc>
+    </group>
```
Remove these as they are subclasses of NXprocess

lukaspie replied:
done

prjemian replied:
For the new added `group` elements, if the `doc` is a repeat of what is in the referenced base class, then the `doc` is not necessary here.

lukaspie replied:
These base classes are now extending `NXprocess` and thus are not in added to `NXprocess` anymore.

phyy-nx commented:
NXcalibration already reviewed
NXdistortion: make a proper subclass of NXprocess, check for duplicated fields and remove them
NXregistration: make a proper subclass of NXprocess, check for duplicated fields and remove them

Accepted by NIAC vote as part of [NIAC2024](https://www.nexusformat.org/NIAC2024.html)

lukaspie commented:
Dev notes to keep track:

- NXdistortion: made a proper subclass of NXprocess, removed duplicated fields
- NXregistration: made a proper subclass of NXprocess, removed duplicated fields
- NXprocess: removed REGISTRATION, CALIBRATION, DISTORTION

depends on / blocked by:

- NXcalibration -> #1419

ToDo:
- rebase after #1419 has been merged

lukaspie commented:
@phyy-nx @sanbrock as `NXcalibration` is now extending `NXprocess`, it is not used _within_ `NXprocess` anymore. Therefore, we could remove it here and bring it in with #1419. In that case, this whole PR comes down to `NXdistortion` and `NXregistration` as sub-classes of `NXprocess`, which has already been voted for by NIAC. So if you agree, I can remove the changes to `NXcalibration` and then this PR should be ready to be finally reviewed. Note that the failing CI/CD also comes from `NXcalibration`, so that would be solved.

phyy-nx commented:
> @phyy-nx @sanbrock as `NXcalibration` is now extending `NXprocess`, it is not used _within_ `NXprocess` anymore. Therefore, we could remove it here and bring it in with #1419. In that case, this whole PR comes down to `NXdistortion` and `NXregistration` as sub-classes of `NXprocess`, which has already been voted for by NIAC. So if you agree, I can remove the changes to `NXcalibration` and then this PR should be ready to be finally reviewed. Note that the failing CI/CD also comes from `NXcalibration`, so that would be solved.

Hm, the comment above says 

> NXcalibration already reviewed

Where was it reviewed?  Also, #1419 has not been voted on but this has.  Wouldn't keeping `NXcalibration` in this PR simplify things?

lukaspie commented:
> > @phyy-nx @sanbrock as `NXcalibration` is now extending `NXprocess`, it is not used _within_ `NXprocess` anymore. Therefore, we could remove it here and bring it in with #1419. In that case, this whole PR comes down to `NXdistortion` and `NXregistration` as sub-classes of `NXprocess`, which has already been voted for by NIAC. So if you agree, I can remove the changes to `NXcalibration` and then this PR should be ready to be finally reviewed. Note that the failing CI/CD also comes from `NXcalibration`, so that would be solved.
> 
> Hm, the comment above says
> 
> > NXcalibration already reviewed
> 
> Where was it reviewed? Also, #1419 has not been voted on but this has. Wouldn't keeping `NXcalibration` in this PR simplify things?

I think #1419 was voted on here: https://github.com/nexusformat/definitions/issues/1452#issuecomment-2382668722 or am I missing something? For me, it would be okay to bring it in here or in #1419, as you prefer. I just didn't want to have to update it in multiple places.

lukaspie commented:
Update: I fixed the build for `NXcalibration`. On second though, it might actually make sense to merge `NXcalibration` with this PR since this is also a subclass for `NXprocess`. The commits are the same, so it does not really matter. In any case, any merging of `NXcalibration` is still blocked by #1486.

phyy-nx commented:
From Telco: subclass NXregistration from NXprocess, and remove NXcalibration (already part of #1419).  Then since this has been voted on, it needs an approval and can be merged.

lukaspie commented:
> From Telco: subclass NXregistration from NXprocess, and remove NXcalibration (already part of #1419). Then since this has been voted on, it needs an approval and can be merged.

@phyy-nx: NXregistration is now subclassed from NXprocess, and I removed NXcalibration. Should be ready for final review and merge. Thanks!

phyy-nx commented:
@lukaspie ok to merge!
